### PR TITLE
Add Testcontainers Service Connection support for Cosmos, Blob and Queue

### PR DESCRIPTION
**Pull request by eddumelendez**
# Description

`ConnectionDetails` and `Service Connection` support have been added for Cosmos, Storage Blob and Storage Queue.

Currently, Cosmos can be used with Testcontainers only using Gateway Mode. See https://github.com/testcontainers/testcontainers-java/issues/5518

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

Fixes spring-projects/spring-boot#40329

### CODE REVIEW

**Review by rujche — APPROVED** [review]

File: sdk/spring/spring-cloud-azure-autoconfigure/src/test/java/com/azure/spring/cloud/autoconfigure/implementation/storage/blob/AzureStorageBlobAutoConfigurationTests.java [hunk 1]
```diff
@@ -230,6 +230,46 @@ void configurationPropertiesShouldBind() {
...
+            .withBean(BlobAsyncClient.class, () -> mock(BlobAsyncClient.class))
+            .withBean(BlobClient.class, () -> mock(BlobClient.class))
+            .withBean(AzureStorageBlobConnectionDetails.class, () -> new CustomAzureStorageBlobConnectionDetails(connectionString))
+            .run(context -> {
+                assertThat(context).hasSingleBean(AzureStorageBlobProperties.class);
+                AzureStorageBlobProperties properties = context.getBean(AzureStorageBlobProperties.class);
+                assertEquals(endpoint, properties.getEndpoint());
+                assertEquals("test-key", properties.getAccountKey());
+                assertEquals("test-sas-token", properties.getSasToken());
+                assertEquals(connectionString, properties.getConnectionString());
```
Comment (rujche) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/test/java/com/azure/spring/cloud/autoconfigure/implementation/storage/blob/AzureStorageBlobAutoConfigurationTests.java:263]: Delete other line except this line to make the test purpose more clear.

File: sdk/spring/spring-cloud-azure-autoconfigure/src/test/java/com/azure/spring/cloud/autoconfigure/implementation/storage/blob/AzureStorageBlobAutoConfigurationTests.java [hunk 2]
```diff
@@ -230,6 +230,46 @@ void configurationPropertiesShouldBind() {
...
+    void connectionDetailsShouldBind() {
+        String accountName = "test-account-name";
+        String connectionString = String.format(STORAGE_CONNECTION_STRING_PATTERN, accountName, "test-key");
+        String endpoint = String.format("https://%s.blob.core.windows.net", accountName);
+        String customerProvidedKey = "fakekey";
+        this.contextRunner
+            .withPropertyValues(
+                "spring.cloud.azure.storage.blob.endpoint=" + endpoint,
+                "spring.cloud.azure.storage.blob.account-key=test-key",
+                "spring.cloud.azure.storage.blob.sas-token=test-sas-token",
```
Comment (rujche) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/test/java/com/azure/spring/cloud/autoconfigure/implementation/storage/blob/AzureStorageBlobAutoConfigurationTests.java:243]: Add `spring.cloud.azure.storage.blob.connection-string=another-value` to confirm that `AzureStorageBlobConnectionDetails` has higher priority.

File: sdk/spring/spring-cloud-azure-testcontainers/src/test/java/com/azure/spring/testcontainers/service/connection/storage/StorageQueueContainerConnectionDetailsFactoryTests.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+import org.springframework.test.context.TestPropertySource;
+import org.springframework.test.context.junit.jupiter.SpringJUnitConfig;
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.junit.jupiter.Container;
+import org.testcontainers.junit.jupiter.Testcontainers;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@SpringJUnitConfig
+@TestPropertySource(properties = "spring.cloud.azure.storage.queue.queue-name=devstoreaccount1/tc-queue")
```
Comment (rujche) [inline sdk/spring/spring-cloud-azure-testcontainers/src/test/java/com/azure/spring/testcontainers/service/connection/storage/StorageQueueContainerConnectionDetailsFactoryTests.java:23]: Keep same style with queue-name and `CosmosContainerConnectionDetails#getDatabase`

**Comment by rujche** [comment]
FYI: About the test failure, I create a issue in spring-boot repo: https://github.com/spring-projects/spring-boot/issues/41100

And for current PR, I'll fix this test failure by a workaround.

**Comment by rujche** [comment]
Hi, @saragluna , @Netyyyy .
**ConnectionDetails** is supported since **spring-boot:3.1**, so **java - spring - compatibility - tests** with **spring-boot:3.0** failed.
One solution is claim that next release of **Spring Cloud Azure** not support **spring-boot:3.0**.
Do you have any other solutions?

File: sdk/spring/spring-cloud-azure-autoconfigure/src/test/java/com/azure/spring/cloud/autoconfigure/implementation/cosmos/AzureCosmosAutoConfigurationTests.java [hunk 4]
```diff
@@ -236,6 +235,30 @@ void configurationPropertiesShouldBind() {
             });
     }
 
+    @Test
+    void connectionDetailsShouldHasHigherPriorityThanProperties() {
```
Comment (rujche) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/test/java/com/azure/spring/cloud/autoconfigure/implementation/cosmos/AzureCosmosAutoConfigurationTests.java:239]: TODO: Add similar logic to other blob and queue.

File: sdk/spring/spring-cloud-azure-testcontainers/src/test/java/com/azure/spring/testcontainers/service/connection/storage/StorageBlobContainerConnectionDetailsFactoryTests.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+@SpringJUnitConfig
+@Testcontainers
+class StorageBlobContainerConnectionDetailsFactoryTests {
+    @Container
+    @ServiceConnection
+    private static final GenericContainer<?> azurite = new GenericContainer<>(
+        "mcr.microsoft.com/azure-storage/azurite:latest")
+        .withExposedPorts(10000)
+        .withCommand("azurite --skipApiVersionCheck && azurite -l /data --blobHost 0.0.0.0 --queueHost 0.0.0.0 --tableHost 0.0.0.0");
```
[the remark this answers is not in the document]
Comment (rujche) [reply sdk/spring/spring-cloud-azure-testcontainers/src/test/java/com/azure/spring/testcontainers/service/connection/storage/StorageBlobContainerConnectionDetailsFactoryTests.java:36]: FYI: `azurite -l /data --blobHost 0.0.0.0 --queueHost 0.0.0.0 --tableHost 0.0.0.0` is original `CMD`.
Screenshot from docker desktop:
> image

File: sdk/spring/spring-cloud-azure-testcontainers/src/test/java/com/azure/spring/cloud/testcontainers/implementation/service/connection/storage/StorageBlobContainerConnectionDetailsFactoryTests.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+@SpringJUnitConfig
+@Testcontainers
+class StorageBlobContainerConnectionDetailsFactoryTests {
+    @Container
+    @ServiceConnection
+    private static final GenericContainer<?> azurite = new GenericContainer<>(
+        "mcr.microsoft.com/azure-storage/azurite:latest")
+        .withExposedPorts(10000)
+        .withCommand("azurite --skipApiVersionCheck && azurite -l /data --blobHost 0.0.0.0 --queueHost 0.0.0.0 --tableHost 0.0.0.0");
```
Comment (rujche) [inline sdk/spring/spring-cloud-azure-testcontainers/src/test/java/com/azure/spring/cloud/testcontainers/implementation/service/connection/storage/StorageBlobContainerConnectionDetailsFactoryTests.java:36]: FYI: `azurite --skipApiVersionCheck` is used to fix pipeline failure.
Refs:
1. Screenshot of pipeline:
> image
2. Doc: https://learn.microsoft.com/en-us/azure/storage/common/storage-use-azurite?tabs=docker-hub%2Cblob-storage#skip-api-version-check

Comment (rujche) [reply sdk/spring/spring-cloud-azure-testcontainers/src/test/java/com/azure/spring/cloud/testcontainers/implementation/service/connection/storage/StorageBlobContainerConnectionDetailsFactoryTests.java:36]: FYI: `azurite -l /data --blobHost 0.0.0.0 --queueHost 0.0.0.0 --tableHost 0.0.0.0` is original `CMD`.
Screenshot from docker desktop:
> image

File: sdk/spring/spring-cloud-azure-testcontainers/src/main/java/com/azure/spring/cloud/testcontainers/implementation/service/connection/storage/StorageQueueContainerConnectionDetailsFactory.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        implements AzureStorageQueueConnectionDetails {
+
+        StorageQueueContainerConnectionDetails(ContainerConnectionSource<Container<?>> source) {
+            super(source);
+        }
+
+        @Override
+        public String getConnectionString() {
+            return "DefaultEndpointsProtocol=http;AccountName=devstoreaccount1;AccountKey=Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==;QueueEndpoint=http://%s:%d/devstoreaccount1;"
+                .formatted(getContainer().getHost(), getContainer().getMappedPort(STORAGE_QUEUE_PORT));
```
Comment (rujche) [inline sdk/spring/spring-cloud-azure-testcontainers/src/main/java/com/azure/spring/cloud/testcontainers/implementation/service/connection/storage/StorageQueueContainerConnectionDetailsFactory.java:39]: FYI: https://learn.microsoft.com/en-us/azure/storage/common/storage-use-azurite?tabs=docker-hub%2Cblob-storage#http-connection-strings

File: sdk/spring/spring-cloud-azure-testcontainers/src/test/java/com/azure/spring/cloud/testcontainers/implementation/service/connection/cosmos/CosmosContainerConnectionDetailsFactoryTests.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+@Testcontainers
+class CosmosContainerConnectionDetailsFactoryTests {
+
+    @TempDir
+    private static File tempFolder;
+
+    @Container
+    @ServiceConnection
+    private static final CosmosDBEmulatorContainer cosmos = new CosmosDBEmulatorContainer(DockerImageName.parse("mcr.microsoft.com/cosmosdb/linux/azure-cosmos-emulator:latest"))
+        .waitingFor(Wait.forHttps("/_explorer/emulator.pem").forStatusCode(200).allowInsecure())
```
Comment (rujche) [inline sdk/spring/spring-cloud-azure-testcontainers/src/test/java/com/azure/spring/cloud/testcontainers/implementation/service/connection/cosmos/CosmosContainerConnectionDetailsFactoryTests.java:49]: Refs: https://learn.microsoft.com/en-us/azure/cosmos-db/how-to-develop-emulator?tabs=docker-windows%2Ccsharp&pivots=api-nosql#import-the-emulators-tlsssl-certificate

**Comment by rujche** [comment]
Hi, @saragluna 

1. About the **java - spring - compatibility-tests**  pipeline failure. It's not possible to make pipeline passed without updating the pipeline's configuration. When the pipeline use **spring-boot:3.0** to build, and **spring-boot:3.0** does not have class like **ConnectionDetails**. **ConnectionDetails** is added since **spring-boot:3.1.0**. So build with **spring-boot:3.0**will always have error like **ClassNotFound**.
2. According to previous point, How to fix pipeline failure? I see that the pipeline set spring-boot:3.0 as test target version because of this file: https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/spring/pipeline/spring-cloud-azure-supported-spring.json. Maybe I should edit this file by creating a PR. But I'm afraid this file will be updated by some job like GitHub actions. Do you have any other options?
3. You mentioned that use some method like **@ConditionalOnClass** to make code compatible with **spring-boot:3.0**. It's not a good idea because:
3.1. Do like this can **NOT** make pipeline pass. Maybe user will not have problem, but the pipeline will always failed, because it can **NOT** find  **ConnectionDetails** when using **spring-boot:3.0**.
3.2. According to previous point (3.1), without pipeline pass, we are not sure that our code compatible with spring-boot:3.0.
3.3. **ConnectionDetails** related logic is coupled with current logic of **XxxAutoConfiguration**. Because **ConnectionDetails** provided connection information has higher priority than properties. It's not easy to split them. 
3.4. Use **@ConditionalOnClass** to solve this problem requires to make **ConnectionDetails** to be a inner class inside a class that annotated with **@ConditionalOnClass**. And all codes may cause **ClassNotFound** should put into the class annotated with **@ConditionalOnClass**, do like this will make a mess of logic, and hard to maintain.
3.5. It's highly possible that in future, **spring-cloud-azure** imported class released since **spring-boot:3.x (x > 0)**. **@ConditionalOnClass** is not silver bullet to solve this kind of problem.

**Comment by rujche** [comment]
Hi, @saragluna 

I think it's time to prepare the document and sample. Now where should I put the docs and samples for **spring-cloud-azure:5.14** which is compatible for **spring-boot:3**?

**Comment by rujche** [comment]
Now all tests passed except compatibility test for spring-boot:3.0.

**Comment by rujche** [comment]
Hi, @saragluna , could you please help to review this PR?

**Comment by eddumelendez** [comment]
Hi @rujche, thanks for taking care of the PR! Looking forward to see it integrated 😃

**Comment by rujche** [comment]
Hi, @saragluna .
Please help to review this PR.

**Comment by rujche** [comment]
Closing in favor of https://github.com/Azure/azure-sdk-for-java/pull/41176

**Pull request closed without merging**